### PR TITLE
[depends] expat 2.2.5, ccache 3.4.1, miniupnpc 2.0.20180203

### PR DESCRIPTION
**Pull request by fanquake**
miniupnpc changelog: http://miniupnp.free.fr/files/changelog.php?file=miniupnpc-2.0.20180203.tar.gz
2.0.20180203 includes fixes for the recent buffer overflow and segfault issues, see https://github.com/miniupnp/miniupnp/issues/268.

expat changelog: https://github.com/libexpat/libexpat/blob/R_2_2_5/expat/Changes
2.2.2 & 2.2.3 included security fixes.

ccache changelog: https://ccache.samba.org/releasenotes.html#_ccache_3_4_1

Also includes latest config.guess and config.sub.

### CODE REVIEW

**Comment by fanquake** [comment]
miniupnpc build error on osx:
```
minissdpc.c:632:35: error: use of undeclared identifier 'IP_MULTICAST_TTL'
                if(setsockopt(sudp, IPPROTO_IP, IP_MULTICAST_TTL, &ttl, ...
                                                ^
minissdpc.c:660:23: error: use of undeclared identifier 'INADDR_NONE'
                        if(mc_if.s_addr != INADDR_NONE)
                                           ^
minissdpc.c:663:37: error: use of undeclared identifier 'IP_MULTICAST_IF'
                                if(setsockopt(sudp, IPPROTO_IP, IP_MULTI...
                                                                ^
minissdpc.c:678:18: error: variable has incomplete type 'struct ifreq'
                                struct ifreq ifr;
                                             ^
minissdpc.c:678:12: note: forward declaration of 'struct ifreq'
                                struct ifreq ifr;
                                       ^
minissdpc.c:680:40: error: use of undeclared identifier 'IFNAMSIZ'
                                strncpy(ifr.ifr_name, multicastif, IFNAMSIZ);
                                                                   ^
/home/travis/build/bitcoin/bitcoin/depends/SDKs/MacOSX10.11.sdk/usr/include/secure/_string.h:119:39: note: 
      expanded from macro 'strncpy'
  __builtin___strncpy_chk (dest, src, len, __darwin_obsz (dest))
                                      ^
minissdpc.c:681:18: error: use of undeclared identifier 'IFNAMSIZ'
                                ifr.ifr_name[IFNAMSIZ-1] = '\0';
                                             ^
minissdpc.c:682:20: error: invalid application of 'sizeof' to an incomplete type
      'struct ifreq'
                                if(ioctl(sudp, SIOCGIFADDR, &ifr, &ifrlen) < 0)
                                               ^~~~~~~~~~~
/home/travis/build/bitcoin/bitcoin/depends/SDKs/MacOSX10.11.sdk/usr/include/sys/sockio.h:94:21: note: 
      expanded from macro 'SIOCGIFADDR'
#define SIOCGIFADDR     _IOWR('i', 33, struct ifreq)    /* get ifnet address */
                        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/travis/build/bitcoin/bitcoin/depends/SDKs/MacOSX10.11.sdk/usr/include/sys/ioccom.h:97:48: note: 
      expanded from macro '_IOWR'
#define _IOWR(g,n,t)    _IOC(IOC_INOUT, (g), (n), sizeof(t))
                        ~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~
/home/travis/build/bitcoin/bitcoin/depends/SDKs/MacOSX10.11.sdk/usr/include/sys/ioccom.h:92:13: note: 
      expanded from macro '_IOC'
        (inout | ((len & IOCPARM_MASK) << 16) | ((group) << 8) | (num))
                   ^
minissdpc.c:678:12: note: forward declaration of 'struct ifreq'
                                struct ifreq ifr;
                                       ^
minissdpc.c:687:37: error: use of undeclared identifier 'IP_MULTICAST_IF'
                                if(setsockopt(sudp, IPPROTO_IP, IP_MULTI...
```

**Comment by fanquake** [comment]
Similar (now solved) issue when building on FreeBSD here: https://github.com/miniupnp/miniupnp/issues/151

**Comment by practicalswift** [comment]
Linking in #12395 for reference.

**Comment by jonasschnelli** [comment]
Gitian depends complication seems to fail (we compile the dependency but don't use it AFAIK):
https://bitcoin.jonasschnelli.ch/build/494

**Comment by MarcoFalke** [comment]
gitian fails due to #12386

**Comment by laanwj** [comment]
> gitian fails due to #12386

That should be solved, but still travis errors, apparently in the macos build.
```
minissdpc.c:632:35: error: use of undeclared identifier 'IP_MULTICAST_TTL'
                if(setsockopt(sudp, IPPROTO_IP, IP_MULTICAST_TTL, &ttl, ...
                                                ^
```

**Comment by fanquake** [comment]
Added a commit that should fix the miniupnpc build issues. I might also pull #11995 in here, so it can be reviewed and merged.

**Comment by laanwj** [comment]
utACK https://github.com/bitcoin/bitcoin/pull/12402/commits/25409b16fd593f45168e839fe21ed0f642908e11

**Comment by theuni** [comment]
The miniupnpc patch here isn't quite right. The _DARWIN_C_SOURCE needs to be defined for darwin only, not universally as it is now.

**Comment by achow101** [comment]
It looks like ccache's bundled zlib is failing to build with mingw which is causing #12515. This was also reported in IRC as causing an issue with a new gitian setup.

**Comment by fanquake** [comment]
This should be addressed in ccache 3.4.2: https://github.com/ccache/ccache/commit/418c419c6bcd6e43eeed2eae5b861441192ebc9e

File: doc/dependencies.md [hunk 1]
```diff
@@ -7,18 +7,18 @@ These are the dependencies currently used by Bitcoin Core. You can find instruct
 | --- | --- | --- | --- | --- | --- |
 | Berkeley DB | [4.8.30](http://www.oracle.com/technetwork/database/database-technologies/berkeleydb/downloads/index.html) | 4.8.x | No |  |  |
 | Boost | [1.64.0](http://www.boost.org/users/download/) | [1.47.0](https://github.com/bitcoin/bitcoin/pull/8920) | No |  |  |
-| ccache | [3.3.4](https://ccache.samba.org/download.html) |  | No |  |  |
+| ccache | [3.3.6](https://ccache.samba.org/download.html) |  | No |  |  |
```
Comment (MarcoFalke) [inline doc/dependencies.md]: Should say 3.4.1?

**Comment by droark** [comment]
Any chance this can get ported over to the 0.16 branch? I'm trying to do Gitian builds right now. Maybe it's a coincidence but Sourceforge chokes whenever Gitian attempts to download expat.

**Comment by fanquake** [comment]
@droark See #12585

**Pull request merged**